target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh

```diff
@@ -8,8 +8,13 @@ preinit_set_mac_address() {
 		ip link set dev eth1 address $(macaddr_add "$base_mac" 3)
 		;;
 	asus,rt-ac42u)
-		ip link set dev eth0 address $(mtd_get_mac_binary_ubi Factory 0x1006)
-		ip link set dev eth1 address $(mtd_get_mac_binary_ubi Factory 0x9006)
+		base_mac=$(mtd_get_mac_binary_ubi Factory 0x1006)
+		ip link set dev eth0 address $base_mac
+		ip link set dev lan1 address $base_mac
+		ip link set dev lan2 address $base_mac
+		ip link set dev lan3 address $base_mac
+		ip link set dev lan4 address $base_mac
```

Comment:
I think if you remove these lan1-4 macaddr assignments, they will still use the macaddr of the DSA parent `eth0`

Reply:
@john-tho  not really true. they would be random MAC when eth0 up.
I have to change mac in preinit for eth0
but at this time the dsa lan1-4 already created and use the parent-random mac.
so I have to change them here.

Reply:
Of course, sorry.
Alternatively, u-boot may be able to set these addresses through `ethernetN` aliases in DTS
I cannot find an OEM firmware image for this device to check if it's u-boot has this setting enabled (`FDT_SEQ_MACADDR_FROM_ENV`), see https://gitlab.com/u-boot/u-boot/-/blob/master/common/fdt_support.c#L625
Example:
https://github.com/openwrt/openwrt/blob/c27b43956407f3adc3cc2693792acd6b40a01877/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ecw5211.dts#L17
https://github.com/openwrt/openwrt/blob/c27b43956407f3adc3cc2693792acd6b40a01877/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-ecw5211.dts#L18

Reply:
QCA has MAC fixup enabled by default, so give aliases a go instead of having to do it from userspace as I dont think NVMEM will work with UBI.

Reply:
@robimarko  how to do that?
mac is lay on the ubi volume "Factory", so how to use nvmem?

Reply:
No, dont use NVMEM, just read what @john-tho wrote and add aliases

Reply:
@robimarko  that would no help, I think.

Reply:
Why?
Try it, by default bootloader will use those aliases to populate MAC in DTS and there is no need to set it from userspace

Reply:
@robimarko  There is not mec in uboot-env, so the uboot can't find the mac and help to fixup the mac by fdt-fixup

Reply:
That's a different situation then, no way around hacking it in userspace

Reply:
it seems that you don't know how uboot fdt-fixup works
if not mac how can uboot set a mac for dts

Reply:
is there anyway to fetch the mac and setup for eth0 and lan1-4?
show me the code

Reply:
I am going to ignore the last 2 replies, since there is no other way what you are doing is fine.